src/Neo.VM/JumpTable/JumpTable.Compound.cs

```diff
@@ -438,19 +440,38 @@ public virtual void SetItem(ExecutionEngine engine, Instruction instruction)
             if (value is Struct s) value = s.Clone(engine.Limits);
             var key = engine.Pop<PrimitiveType>();
             var x = engine.Pop();
+            var isRC2 = engine.ReferenceCounter.Version == RCVersion.V2;
             switch (x)
             {
                 case VMArray array:
                     {
                         var index = (int)key.GetInteger();
                         if (index < 0 || index >= array.Count)
                             throw new CatchableException($"The index of {nameof(VMArray)} is out of range, {index}/[0, {array.Count}).");
+                        if (isRC2)
+                            engine.ReferenceCounter.RemoveStackReference(array[index]);
                         array[index] = value;
+                        if (isRC2)
+                            engine.ReferenceCounter.AddStackReference(value);
                         break;
                     }
                 case Map map:
                     {
+                        if (isRC2)
+                        {
+                            if (!map.TryGetValue(key, out var value1))
+                            {
+                                engine.ReferenceCounter.AddStackReference(key);
```

Comment:
the key is added multiple times? @roman-khimov

Reply:
no, `TryGetValue` 1st, doesn't exist add it.

Reply:
the key is added if it does not exsit in the dictionary,   when the value is removed from the dic, the key is removed as well. So no value also means no key.

Reply:
But if i add twice the key is only one and is counted as two

Reply:
if you add twice, the key wont be changed, key will only be added once, then it will only be removing the value, and add the value. 

The complete logic here is:

key exists:

         remove old value ref, add new value ref
         
key does not exist:
         add key ref, add new value ref.

Reply:
The key don't change in the map, but it's called twice to `engine.ReferenceCounter.AddStackReference(key);`

Reply:
> the key is added multiple times? @roman-khimov

No, the key is added only once. This code is correct.

>  but it's called twice to engine.ReferenceCounter.AddStackReference(key);

This code adds key ref only if the key is not yet in the map. So if it's the new key, then it's added to the RC. But if this key is already exist in the map, then no need to add it to RC one more time (only value should be added in this case).